Lib/dis.py

```diff
@@ -467,56 +380,199 @@ def is_jump_target(self):
         """True if other code jumps to here, otherwise False"""
         return self.label is not None
 
-    def _disassemble(self, lineno_width=3, mark_as_current=False, offset_width=0,
-                           label_width=0):
-        """Format instruction details for inclusion in disassembly output.
+    def __str__(self):
+        output = io.StringIO()
+        formatter = Formatter(file=output,
+                              lineno_width=3,
+                              offset_width=0,
+                              label_width=0)
+        formatter.print_instruction(self, False)
+        return output.getvalue()
+
+
+class Formatter:
+
+    NO_LINENO = '  --'
+
+    def __init__(self, file=None, lineno_width=3, offset_width=0, label_width=0,
+                       linestarts=None, exception_entries=None, line_offset=0):
+        """Create a Formatter
 
+        *file* where to write the output
         *lineno_width* sets the width of the line number field (0 omits it)
-        *mark_as_current* inserts a '-->' marker arrow as part of the line
         *offset_width* sets the width of the instruction offset field
         *label_width* sets the width of the label field
+
+        *linestarts* dictionary mapping offset to lineno, for offsets that
+                     start a new line
+        *exception_entries* the exception table (list of ExceptionEntry)
+        *line_offset* the line number (within the code unit)
         """
+        self.file = file
+        self.offset_width = offset_width
+        self.label_width = label_width
+        self.linestarts = linestarts
```

Comment:
It uses `linestarts` to figure out the size of the line number field.

The formatter _is_ for a sequence of instructions. That's the idea behind moving the formatting code off the Instruction class into a separate class. Actually even the Instructions have data on them that ties them to a particular sequence (like the offset).

Reply:
The issue here really is that the formatter works on an instructions iterable (which is sometimes a generator) so it can only traverse it once. So it can't do anything like "find the max label/line number/offset" before it begins producing output.

Reply:
The name "formatter" implies to me that it just does formatting.
I don't have a better name of the top of my head, but add a comment/docstring explaining this